swim/handlers_test.go

```diff
@@ -0,0 +1,205 @@
+	s.Require().False(s.testNode.node.gossip.Stopped())
+
+	_, err := s.testNode.node.gossipHandlerStop(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+
+	s.True(s.testNode.node.gossip.Stopped())
+}
+
+func (s *HandlerTestSuite) TestToggleGossipHandler() {
+	s.Require().True(s.testNode.node.gossip.Stopped())
+
+	var err error
+
+	// Toggle gossip on
+	_, err = s.testNode.node.gossipHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.False(s.testNode.node.gossip.Stopped())
+
+	// Toggle gossip off
+	_, err = s.testNode.node.gossipHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.True(s.testNode.node.gossip.Stopped())
+}
+
+func (s *HandlerTestSuite) TestAdminLeaveJoinHandlers() {
+	// Join test node to cluster
+	s.cluster.Add(s.testNode.node)
+	s.Require().Equal(5, s.testNode.node.CountReachableMembers(), "expect a cluster of 5")
+
+	var err error
+	var status *Status
+
+	// Test leave handler works correctly
+	status, err = s.testNode.node.adminLeaveHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.Equal(&Status{Status: "ok"}, status)
+	s.Equal(4, s.testNode.node.CountReachableMembers())
+
+	// Test join handler brings it back to 5
+	status, err = s.testNode.node.adminJoinHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.Equal(&Status{Status: "rejoined"}, status)
+	s.Equal(5, s.testNode.node.CountReachableMembers())
+}
+
+// TestRegisterHandlers tests that registerHandler always succeeds.
+func (s *HandlerTestSuite) TestRegisterHandlers() {
+	s.NoError(s.testNode.node.registerHandlers())
+}
+
+func (s *HandlerTestSuite) TestNotImplementedHandler() {
+	_, err := notImplementedHandler(s.ctx, &arg{})
+	s.EqualError(err, "handler not implemented")
+}
+
+// TestErrorHandler tests that the errorHandler logs the correct error message.
+func (s *HandlerTestSuite) TestErrorHandler() {
+	logger := &mocks.Logger{}
+
+	errTest := errors.New("test error")
+
+	logger.On("WithField", "error", errTest).Return(logger)
+	logger.On("Info", []interface{}{"error occurred"})
+
+	originalLogger := s.testNode.node.log
+	s.testNode.node.log = logger
+
+	s.testNode.node.errorHandler(s.ctx, errTest)
+	logger.AssertExpectations(s.T())
+
+	// Restore old logger, as stuff will be logged on teardown
+	s.testNode.node.log = originalLogger
+}
+
+func (s *HandlerTestSuite) TestTickHandler() {
+	s.testNode.node.Stop()
+
+	ping, err := s.testNode.node.tickHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+
+	s.Equal(s.testNode.node.memberlist.Checksum(), ping.Checksum, "checksum should be returned by tickHandler")
+}
+
+// TestJoinHandlerError tests that the join handler returns an error if the
+// join request is invalid.
+func (s *HandlerTestSuite) TestJoinHandlerError() {
+	// Construct an invalid joinRequest
+	req := &joinRequest{}
```

Comment:
Generally speaking, I think we need to do a better job of building self-documenting fixtures. This is a place where I feel a function or fixture can exist that explicitly defines an "invalid request" rather than the comment stating so and the line beneath it. I'm not sure what idiomatic fixtures in Go look like though. So let's postpone and discuss at a later date.